### PR TITLE
Add coverall support for continuous code coverage testing

### LINKED ISSUE #993: Automate Code Coverage Testing
pg_shard and [cstore_fdw](https://coveralls.io/github/citusdata/cstore_fdw) already use coveralls.io, although maybe codecov [is the new hotness](https://www.slant.co/versus/7928/7929/~coveralls_vs_codecov).

### PR DESCRIPTION
**Pull request by lithp**
Fixes #993, code coverage results show up here: https://coveralls.io/github/citusdata/citus

A part of the #890 epic.

Also uploads coverage data to codecov, which can be viewed here: https://codecov.io/gh/citusdata/citus

### CODE REVIEW

**Comment by onderkalaci** [comment]
This is cool! 
With `pg_shard`, I remember that we were not allowed to check-in any PRs if it decreases the code coverage. Are we going to have that here as well?

**Comment by lithp** [comment]
@onderkalaci I don't know. I'm a little afraid of over-indexing on code coverage since it's a sloppy proxy for "have we tested all the things which could lead to bugs". But that still sounds like a reasonable policy, maybe email it to devs@?

Also, @jasonmp85, any chance you could take a look at this?

**Comment by metdos** [comment]
@lithp, as long as I call, this is not one of the high-ranked test automation issues we decided to tackle in our development meeting. Why did we start with this?

If there is not any strong reason, could we pause on this?

**Comment by lithp** [comment]
@metdos Mostly because I believe in do-ocracy but also because this is such a short task, there's no reason to block a task which takes a single day and neatly fits into our goal of making testing easier.

**Comment by jasonmp85** [comment]
> Mostly because I believe in do-ocracy

+:100:

**Comment by metdos** [comment]
@lithp, @jasonmp85, okay if you think that this is not stealing time from what all the developers collectively decided as more important, you can go ahead and merge this. I have no objection for that.

**Comment by sumedhpathak** [comment]
@lithp If its a short task, I am also OK with it but I had a few questions:

1. Does this remove the lineitem we have in our test plan for code coverage? How much time does that save us? And can we track it somewhere that we no longer need to run that during release?
2. When does this new feature run? Does it run after every check-in?
3. How do we interpret/review the results?

**Comment by lithp** [comment]
> Does this remove the lineitem we have in our test plan for code coverage? How much time does that save us? And can we track it somewhere that we no longer need to run that during release?

Yes, this removes the need to manually check code coverage at the end of release cycles. Eren [estimated](https://docs.google.com/a/citusdata.com/spreadsheets/d/1PUPAxxf4wN8n1QoW3ZjqxIrh3ilQ_EXH17k8tuoEW_0/edit?usp=sharing) (link only works for citusdata.com users, sorry) that code coverage testing currently takes us half a day per release cycle. And yes, if this gets merged in I'll make a note on #890 and also watch out during the next cycle that it's not part of the plan.

> When does this new feature run? Does it run after every check-in?

As of this PR Travis now builds citus with coverage flags enabled and uploads the coverage results at the end of every build, we'll be able to compare coverage between any two commits.

> How do we interpret/review the results? (Edit: The starburst was confusing, but after playing around I think I am getting the hang of it)

This question sounds difficult to answer :)

File: Makefile.global.in [hunk 1]
```diff
@@ -17,6 +17,12 @@ postgres_abs_builddir:=@POSTGRES_BUILDDIR@
 PG_CONFIG:=@PG_CONFIG@
 PGXS:=$(shell $(PG_CONFIG) --pgxs)
 
+ifeq ($(enable_coverage),yes)
+	export PG_CPPFLAGS += -ftest-coverage -fprofile-arcs
```
Comment (jasonmp85) [inline Makefile.global.in]: What's the difference between this and the `--coverage` flag used by `pg_shard` or `cstore_fdw`?

File: src/backend/distributed/Makefile [hunk 2]
```diff
@@ -97,7 +97,7 @@ $(EXTENSION)--6.0-18.sql: $(EXTENSION)--6.0-17.sql $(EXTENSION)--6.0-17--6.0-18.
 
 NO_PGXS = 1
 
-SHLIB_LINK = $(libpq)
+SHLIB_LINK += $(libpq)
```
Comment (jasonmp85) [inline src/backend/distributed/Makefile]: Why did this need to change?

File: Makefile.global.in (hunk 1 above)
Comment (jasonmp85) [inline Makefile.global.in]: Figured out this was coming from PostgreSQL's `enable-coverage` flag. I'm OK copying it.

File: src/backend/distributed/Makefile (hunk 2 above)
Comment (jasonmp85) [inline src/backend/distributed/Makefile]: (I figured out why, but this kind of thing might require a comment in the commit message next time)

**Comment by jasonmp85** [comment]
Switched to the configure flag and added support for it to `citusdata/tools`. Merging.

**Comment by lithp** [comment]
Thanks for pushing this through! Agree that your changes are better.

**Pull request merged**